src/commands/mod.rs

```diff
@@ -2005,14 +2008,71 @@ fn cmd_new(ui: &mut Ui, command: &CommandHelper, args: &NewArgs) -> Result<(), C
         "expected a non-empty list from clap"
     );
     let commits = resolve_base_revs(&workspace_command, &args.revisions)?;
-    let parent_ids = commits.iter().map(|c| c.id().clone()).collect();
+    let parent_ids = commits.iter().map(|c| c.id().clone()).collect_vec();
     let mut tx = workspace_command.start_transaction("new empty commit");
-    let merged_tree = merge_commit_trees(tx.base_repo().as_repo_ref(), &commits);
-    let new_commit = tx
-        .mut_repo()
-        .new_commit(command.settings(), parent_ids, merged_tree.id().clone())
-        .set_description(&args.message)
-        .write()?;
+    let mut num_rebased = 0;
+    let new_commit;
+    if args.insert_before {
+        // Instead of having the new commit as a child of the changes given on the
+        // command line, add it between the changes' parents and the changes.
+        // The parent of the new commit will be the parents of the target commits
+        // which are not descendants of other target commits.
+        let child_ids = parent_ids;
+        let children = RevsetExpression::commits(child_ids.clone());
+        let new_parents = children.parents().minus(&children.descendants());
```

Comment:
I think not. For example, starting from

```
    @   F
    |\  
    o | E
    | o D
    |/  
    | o C
    | o B
    | o A
    |/  
    o root
```

If we want to insert a commit `before children=[B,D]`, the expression above returns `[root,A]` while `children.parents().roots()` returns `root`.

However, this makes me think that there is no good reason not to base the new commit on `A` alone instead of refusing to merge `root` and `A`, I'll add this.

Reply:
Thanks for providing an example. I didn't understand what the goal was before. Maybe `children().parents().heads()` is what we want then? I think that's the generic version of your check for the root commit.

Actually, `jj new A B` (without `--insert-before`) currently doesn't filter out ancestors, so it creates a merge with A and B as parents. I'm not sure if we should error out in that case, or if we should continue to allow it, or if we should make it result in a new commit on top of just B. Whatever we do, it seems obvious that we should make `--insert-before` should match. So since we currently allow it, I would make `--insert-before` match that. That would mean doing just `let new_parents = children.parents()` here. What do you think? That seems like the simplest thing to do for now, and then we can decide about filtering out ancestors at a later time.

Reply:
However, we can use the following

```rust
        let parents = children.parents();
        let new_parents = parents.minus(&children.dag_range_to(&parents));
```

This starts from the parents and should be more efficient.

Reply:
> Actually, `jj new A B` (without `--insert-before`) currently doesn't filter out ancestors, so it creates a merge with A and B as parents. I'm not sure if we should error out in that case, or if we should continue to allow it, or if we should make it result in a new commit on top of just B.

For me, saying `jj new A B` means that you really want a merge commit whose parents are exactly A and B.

> Whatever we do, it seems obvious that we should make `--insert-before` should match. So since we currently allow it, I would make `--insert-before` match that. That would mean doing just `let new_parents = children.parents()` here. What do you think? That seems like the simplest thing to do for now, and then we can decide about filtering out ancestors at a later time.

Saying `jj new --insert-before A B` will make one new commit whose parents are exactly A and C, even if A is an ancestor of C, so the situation is symetrical. What we filter out are the parents of the new commit: we want them to be the ancestors of A and C, but we cannot use any descendants of A or C since those two commits are children of the new commit.

For example

```
    @ C
    o B
    o A
    |
    o root
```

becomes (`jj new -m D --insert-before A C):

```
    o B
    | o C
    o | A
    |/
    @ D
    o root
```

Reply:
Ah, that's a good point. I think I'd prefer that case to be an error. I'm not sure what I would expect the result to be if we allowed it. I feel like I would expect the first step of the command to create D as a child of root and B, but then the rebase of A on top of it would fail. You decided to filter out B as parent for that reason, but I suspect it's better to error out. I can't really see a use case for the command anyway. What do you think?

Reply:
I see this as analogous to rebasing onto a descendant, where currently fail like this:
```
$ jj rebase -s @-- -d @-
Error: Cannot rebase 8f7bb04045cc onto descendant f412199bbd1f
```

Reply:
Although I would really like to be able to do what you describe (`jj rebase -s @-- -d @-`), I agree that refusing to create a loop is a good idea (opposed to trying to break the loop as I did) as I cannot see any useful use case.

We don't have the same problem with `--insert-after`, as we put all ancestors of our targets before the new node (see `test_new_insert_after_children()`).

Reply:
Done, this is now an error (see `test_new_insert_before_no_loop()`)